tests/baselines/reference/mappedTypeAsClauseRelationships.errors.txt

```diff
@@ -0,0 +1,40 @@
+tests/cases/conformance/types/mapped/mappedTypeAsClauseRelationships.ts(12,9): error TS2322: Type 'T' is not assignable to type 'Modify<T>'.
+tests/cases/conformance/types/mapped/mappedTypeAsClauseRelationships.ts(23,9): error TS2322: Type 'T' is not assignable to type 'FilterExclOpt<T>'.
+tests/cases/conformance/types/mapped/mappedTypeAsClauseRelationships.ts(24,9): error TS2322: Type 'T' is not assignable to type 'ModifyExclOpt<T>'.
+
+
+==== tests/cases/conformance/types/mapped/mappedTypeAsClauseRelationships.ts (3 errors) ====
+    // From original issue #45212:
+    type Methods<T> = { [P in keyof T as T[P] extends Function ? P : never]: T[P] };
+    type H<T> = T[keyof Methods<T>]; // Ok
+    
+    // `Filter<T>` only filters out some keys of `T`.
+    type Filter<T> = { [P in keyof T as T[P] extends Function ? P : never]: T[P] };
+    // `Modify<T>` might modify some keys of `T`.
+    type Modify<T> = { [P in keyof T as P extends string? `bool${P}`: P]: T[P] };
+    
+    function fun<T>(val: T) {
+        let x: Filter<T> = val; // Ok
+        let y: Modify<T> = val; // Error
+            ~
+!!! error TS2322: Type 'T' is not assignable to type 'Modify<T>'.
+    }
+    
+    type FilterInclOpt<T> = { [P in keyof T as T[P] extends Function ? P : never]+?: T[P] };
+    type ModifyInclOpt<T> = { [P in keyof T as P extends string? `bool${P}`: never ]+?: T[P] };
+    type FilterExclOpt<T> = { [P in keyof T as T[P] extends Function ? P : never]-?: T[P] };
+    type ModifyExclOpt<T> = { [P in keyof T as P extends string? `bool${P}`: never ]-?: T[P] };
+    
+    function fun2<T>(val: T) {
+        let x: FilterInclOpt<T> = val; // Ok
+        let y: ModifyInclOpt<T> = val; // Ok

```

Comment:
My initial reaction to this one was that this is wrong because `T` might have some property that would conflict with the remapped property type, e.g. if `T` was instantiated with `{ foo: string, boolfoo: number }`. However, it looks like that’s an unsoundness that [already exists](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChkHIAeA-AFzIDOYUoA5gNz6ECe5VN9ThyAXu9VohGuAL65cMAK4gEYYOhDIYAHgAqAPgAURCmoCUOZgQSLqyAPoUMWALzEmooA) independent of mapped types entirely:

```ts
interface Foo {
    x?: string;
    y?: string;
    z?: string;
}

function f<T>(x: T) {
    const _: Foo = x;
}
```

Reply:
Yeah, I tried to follow the already existing logic and reuse everything that's already there, so there might be inconsistencies that were already there or that result from the kind of types that are in key remapping clauses.

Reply:
Optional properties are a pretty big soundness hole, unfortunately. The unconstrained generic one is actually one I have a PR up to remove - https://github.com/microsoft/TypeScript/pull/33570

Reply:
This exact discussed case has changed behavior in https://github.com/microsoft/TypeScript/commit/51b346d65a7d5e474a40113f72dc66106715a47a#diff-25f367dc11e40edc9febb9d6f088e13f9c7c8086bb8b9f85aed321f0903bc46eR31-R33 

One of my PRs bring back original behavior (somewhat accidentally, I wasn't aiming for it), see [here](https://github.com/microsoft/TypeScript/pull/49064/files#diff-25f367dc11e40edc9febb9d6f088e13f9c7c8086bb8b9f85aed321f0903bc46eL33-L35)

Which one is correct? 😅